zuko/flows/mixture.py

```diff
@@ -7,15 +7,73 @@
 import torch
 import torch.nn as nn
 
-from math import prod
-from torch import Tensor
-from torch.distributions import Distribution, MultivariateNormal
 
 # isort: local
 from .core import LazyDistribution
 from ..distributions import Mixture
 from ..nn import MLP
 from ..utils import unpack
+from math import prod
+from torch import Tensor
+from torch.distributions import (
+    Distribution,
+    Independent,
+    LowRankMultivariateNormal,
+    MultivariateNormal,
+    Normal,
+)
+
+
+def _determine_shapes(components, features, covariance_type, tied, cov_rank):
```

Comment:
The following pattern would allow to reduce code duplication

```python
leading = 1 if tied else components

if covariance_type == 'full':
    shapes.extend([
        (leading, features),
        (leading, features * (features - 1) // 2),
    ])
elif ...
```